aeneas/src/ir/PackingSolver.v3

```diff
@@ -0,0 +1,339 @@
+			assignments[CaseField] = interval;
+			var result = solve(idx + 1, state);
+			if (result != null) return result;
+			state[caseIdx].unassignInterval(interval);
+		}
+		return null;
+	}
+
+	// Find a packing for a single scalar
+	def solveScalar(problem: PackingProblem) -> PackingSolution {
+		cases = problem.cases;
+		var isRefScalar = false;
+		for (i < cases.length) {
+			var caseHasRef = false;
+			for (field in cases[i]) {
+				match (field) {
+					Ref => {
+						if (caseHasRef) return null; // cannot pack more than one ref into a single scalar per case
+						isRefScalar = caseHasRef = true;
+					}
+					_ => ;
+				}
+			}
+		}
+
+		var patterns = Array<PackingPattern>.new(cases.length);
+		if (isRefScalar) {
+			for (i < cases.length) {
+				var c = cases[i];
+				var casePacking: PackingPattern;
+
+				var containsRef = false;
+				for (j < cases[i].length) {
+					var f = cases[i][j], fl = CaseField(i, j);
+					match (f) {
+						Ref => {
+							assignments[fl] = refPatterns.refInterval;
+							casePacking = refPatterns.ptrref.copy();
+							casePacking.assignInterval(refPatterns.refInterval);
+							containsRef = true;
+						}
+						_ => ;
+					}
+				}
+
+				if (!containsRef) casePacking = refPatterns.nonptrref.copy();
+				patterns[i] = casePacking;
+			}
+		} else {
+			// can use all bits
+			for (i < cases.length) patterns[i] = refPatterns.nonref.copy();
+		}
+
+		for (assignment in problem.assignments) {
+			var cf = assignment.0, interval = assignment.1;
+			patterns[cf.caseIdx].assignInterval(interval);
+			assignments[cf] = interval;
+		}
+
+		var fieldsToSolve = Vector<CaseField>.new();
+		for (i < cases.length) {
+			for (j < cases[i].length) {
+				var cf = CaseField(i, j);
+				if (!assignments.has(cf)) fieldsToSolve.put(cf);
+			}
+		}
+		fieldOrder = fieldsToSolve.extract();
+		this.cases = cases;
+		
+		var solution = solve(0, patterns);
+		if (solution == null) return null;
+
+		return PackingSolution.new(patterns, assignments, isRefScalar, problem);
+	}
+}
+
+type PackingProblem(cases: Array<Array<PackingField>>, assignments: Array<(CaseField, Interval)>) #unboxed {
+	def render(sb: StringBuilder) -> StringBuilder {
+		var h = HashMap<CaseField, Interval>.new(CaseField.hash, CaseField.==);
+		for (a in assignments) h[a.0] = a.1;
+
+		for (i < cases.length) {
+			if (i > 0) sb.csp();
+			sb.put1("#%d=[", i);
+			for (j < cases[i].length) {
+				if (j > 0) sb.csp();
+				match (cases[i][j]) {
+					Nonref(s) => sb.put1("%d", s);
+					Ref => sb.puts("ref");
+				}
+				var cf = CaseField(i, j);
+				if (h.has(cf)) sb.put1("@%q", h[cf].render);
+			}
+			sb.puts("]");
+		}
+		return sb;
+	}
+}
+
+class PackingSolution(patterns: Array<PackingPattern>, assignments: HashMap<CaseField, Interval>, isRef: bool, problem: PackingProblem) {}
+
+type RefPatterns(
+	ptrref: PackingPattern,
+	refInterval: Interval,
+	nonptrref: PackingPattern,
+	nonref: PackingPattern
+);
+
+component MachPatterns {
```

Comment:
This should definitely be part of the `Target`.

Reply:
I've kept these values as a component, and set the value of `refPattern` in the constructor in `Linux.v3`. Wdyt?